### PR TITLE
adding support for Matrix3/4

### PR DESCRIPTION
**Pull request by SebastianBaltes**
Hi,

thanks a lot for this great library! I needed to set the "uvTransform" - attribute, which wasn't supported, so I extended the three-instanced-uniforms-mesh to support matrix3/4 (and I guess I also fixed array support). Another change is that I needed to remove previous declarations of uvTransform in the shader (see createInstancedUniformsDerivedMaterial line 30). I would be glad if this PR could be integrated. If there is any problem, please let me know.

Cheers,
Sebastian

### CODE REVIEW

File: packages/three-instanced-uniforms-mesh/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "three-instanced-uniforms-mesh",
-  "version": "0.44.0",
+  "version": "0.44.1",
```
Comment (lojjic) [inline packages/three-instanced-uniforms-mesh/package.json:3]: Please revert this; version bumps are handled by the publishing scripts.

File: packages/three-instanced-uniforms-mesh/src/InstancedUniformsDerivedMaterial.js [hunk 2]
```diff
@@ -24,9 +24,14 @@ export function createInstancedUniformsDerivedMaterial (baseMaterial) {
       _uniformNames.forEach((name) => {
         let vertType = vertexUniforms[name]
         let fragType = fragmentUniforms[name]
-        if (vertType || fragType) {
+        const type = vertType || fragType
+        if (type) {
+          // remove existing deklarations
+          const findDeklaration = new RegExp(`\\s*(?:varying|attribute|uniform)\\s+${type}\\s+${name}\\s*;?`, 'g')
+          vertexShader = vertexShader.replaceAll(findDeklaration,"")
```
Comment (lojjic) [inline packages/three-instanced-uniforms-mesh/src/InstancedUniformsDerivedMaterial.js:31]: This seems fine to me, but out of curiosity why did you need to remove the existing declaration?

Not critical, but deklaration->declaration would be easier on my English-speaking eyes. 😉

File: packages/three-instanced-uniforms-mesh/src/InstancedUniformsMesh.js [hunk 3]
```diff
@@ -122,6 +122,10 @@ function setAttributeValue (attr, index, value) {
     }
   } else if (size === 4) {
     attr.setXYZW(index, value.x, value.y, value.z, value.w)
+  } else if (value?.toArray) {
+    value.toArray(attr, index * size)
```
Comment (lojjic) [inline packages/three-instanced-uniforms-mesh/src/InstancedUniformsMesh.js:126]: I'm not quite understanding how this works, shouldn't this be calling toArray with `attr.array` instead of `attr`?

File: packages/three-instanced-uniforms-mesh/src/InstancedUniformsMesh.js [hunk 4]
```diff
@@ -143,7 +147,8 @@ function getItemSizeForValue (value) {
     : typeof value === 'number' ? 1
     : value.isVector2 ? 2
     : value.isVector3 || value.isColor ? 3
-    : value.isVector4 ? 4
+    : value.isVector4 || value.isQuaternion ? 4
+    : value?.elements ? value.elements.length
```
Comment (lojjic) [inline packages/three-instanced-uniforms-mesh/src/InstancedUniformsMesh.js:151]: Can you remove both instances of the nullish coalescing `?.` in this file? They're redundant here (the code would have already errored if value was nullish) and it tends to add verbose code when transpiled.

**Review by lojjic — CHANGES_REQUESTED** [review]
Very cool, thank you for the enhancement!

At the time I wrote this utility, `mat3`/`mat4` attributes were not supported by Three.js, so this was a surprise to me. I went looking and it seems they indeed [added that support at r132](https://github.com/mrdoob/three.js/pull/16141). 🎉  It may be worth adding a note to the README that those will only work with r132+.

Otherwise, I've got a few initial comments for you to address, and I'll have to test this out to verify it's working, but I'm excited to get this merged in!

File: packages/three-instanced-uniforms-mesh/package.json (hunk 1 above)
Comment (SebastianBaltes) [reply packages/three-instanced-uniforms-mesh/package.json]: done

File: packages/three-instanced-uniforms-mesh/src/InstancedUniformsDerivedMaterial.js (hunk 2 above)
Comment (SebastianBaltes) [reply packages/three-instanced-uniforms-mesh/src/InstancedUniformsDerivedMaterial.js]: I changed it to declaration, and thank you for the correction, my german fingers just seem to like the harder consonants by nature ;)

I need to remove existing declarations for "troika_attr_uvTransform", because otherwise I get an "Vertex shader is not compiled" error because of an "troika_attr_uvTransform' : redefinition", it was allready declared as "uniform mat3 troika_attr_uvTransform;" by threejs and this collides with the new "attribute mat3 troika_attr_uvTransform;":

[shader-error.txt](https://github.com/protectwise/troika/files/7801690/shader-error.txt)

File: packages/three-instanced-uniforms-mesh/src/InstancedUniformsMesh.js (hunk 3 above)
Comment (SebastianBaltes) [reply packages/three-instanced-uniforms-mesh/src/InstancedUniformsMesh.js]: You are right, it needs to be attr.array. Done.

File: packages/three-instanced-uniforms-mesh/src/InstancedUniformsMesh.js (hunk 4 above)
Comment (SebastianBaltes) [reply packages/three-instanced-uniforms-mesh/src/InstancedUniformsMesh.js]: done

**Comment by SebastianBaltes** [comment]
Hi Lojjic, sorry it took some time, I was on holiday

File: packages/three-instanced-uniforms-mesh/src/InstancedUniformsDerivedMaterial.js (hunk 2 above)
Comment (lojjic) [reply packages/three-instanced-uniforms-mesh/src/InstancedUniformsDerivedMaterial.js]: Ah, I see! I think that is caused by it renaming the original uniform declaration when it shouldn't -- at the top of the file there's a `precededByUniformRE` regex that should include mat3/mat4 now, I think that would also fix the issue by preventing the declaration from getting renamed.

Although, I'm thinking your approach may be better anyway, since it doesn't leave unused uniforms in the code ...? In which case we could remove the `precededByUniformRE` checks altogether. 🤔 Let me test this out and I may go ahead and remove that after merging your code.

**Review by lojjic — APPROVED** [review]

**Comment by lojjic** [comment]
Thanks @SebastianBaltes! I tested it and it works great, and I was able to remove that `precededByUniformRE` in favor of your declaration removal. I'll do a 0.46 release this week if all goes to plan.

**Comment by SebastianBaltes** [comment]
That's great, thanks a lot!
